kratos/utilities/geometry_utilities.h

```diff
@@ -451,42 +402,38 @@ class GeometryUtils
      */
     template<std::size_t TSize>
-    static void CalculateTriangleDistancesToLine(Element::GeometryType& ThisGeometry, array_1d<double, TSize>& Distances)
+    static void CalculateExactDistancesToPlane(
+        Element::GeometryType& rThisGeometry,
+        array_1d<double, TSize>& rDistances)
     {
-        array_1d<Point, 4> intersection_points;
-        int number_of_intersection_points = CalculateTetrahedraIntersectionPoints(ThisGeometry, Distances, intersection_points);
+        array_1d<Point, TSize> intersection_points;
+        int number_of_intersection_points = CalculateTetrahedraIntersectionPoints(rThisGeometry, rDistances, intersection_points);
 
 		if(number_of_intersection_points == 0) {
-			KRATOS_WARNING("GeometryUtilities") << "Warning: The intersection with interface hasn't found! The distances are" << Distances << std::endl;
+			KRATOS_WARNING("GeometryUtilities") << "Warning: The intersection with interface hasn't found! The distances are" << rDistances << std::endl;
 		}
-		else if(number_of_intersection_points == 1) { 
-            for(unsigned int i_node = 0; i_node < ThisGeometry.size() ; i_node++) {
-				double distance_to_node = norm_2(intersection_points[0] - ThisGeometry[i_node]);
-				Distances[i_node] = std::min(fabs(Distances[i_node]), distance_to_node);
-			}
-		}
-		else if(number_of_intersection_points == 2) {
-            for(unsigned int i_node = 0; i_node < ThisGeometry.size() ; i_node++) {
-                double x0 = ThisGeometry[i_node].X();   double y0 = ThisGeometry[i_node].Y();
-                double x1 = intersection_points[0].X(); double y1 = intersection_points[0].Y();
-                double x2 = intersection_points[1].X(); double y2 = intersection_points[1].Y();
-                //From https://en.wikipedia.org/wiki/Distance_from_a_point_to_a_line#Line_defined_by_two_points
-                double distance_to_line = fabs((y2-y1)*x0 - (x2-x1)*y0 +x2*y1 -y2*x1) / std::sqrt(std::pow(y2-y1,2) + std::pow(x2-x1,2));
-                Distances[i_node] = std::min(fabs(Distances[i_node]), distance_to_line);
-			}
-		}
-		else {
-            KRATOS_WARNING("GeometryUtilities") << "Warning: This is a triangle with more than two intersections! The distances are: " << Distances << std::endl;
-		}
-    }
+        else {
+            BoundedMatrix<double,TSize,TSize-1> DN_DX;
+            array_1d<double, TSize> N;
+            double volume;
+            GeometryUtils::CalculateGeometryData(rThisGeometry, DN_DX, N, volume);
+            array_1d<double, TSize-1> distance_gradient = prod(trans(DN_DX), rDistances);
+            distance_gradient /= norm_2(distance_gradient);
+            #pragma omp parallel for
+            for (int i = 0; i < static_cast<int>(TSize); i++) {
+                //We use the first intersection point as reference
+                double d = std::abs(inner_prod(rThisGeometry[i].Coordinates()-intersection_points[0], distance_gradient));
```

Comment:
I am sorry in case I am saying something stupid but I made the the following observation:
The latest changes were copied into a local branch and compiled in FullDebug mode. The error `Check failed in file /home/simon/Library/boost_1_68_0/boost/numeric/ublas/vector_expression.hpp at line 1677:
e1_.size () == e2_.size ()` appeared and I think it originates here. As I could follow the code, `rThisGeometry[i].Coordinates()-intersection_points[0]` has 3 entries. In contrast `distance_gradient` has only 2 entries in a 2D case. No problems appear in 3D applications. 

KRATOS_WATCH confirmed the described situation for 2D:
rThisGeometry[i].Coordinates() : [3] [ 0,0,0 ]
intersection_points[0] : Point[ 0.125, 0.25, 0 ]
distance_gradient : [2][ -5.55112e-17,1 ]

Reply:
I think that @swenczowski is completely right. Find a suggestion below.